### PR TITLE
8271589: fatal error with variable shift count integer rotate operation.

### PR DESCRIPTION
**Pull request by jatin-bhateja**
Problem seen on targets which do not support variable vector shifts, in such a case vector rotation node inferred by auto-vectorizer are not disintegrable into LeftShift/RightShift and Or operations.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8271589](https://bugs.openjdk.java.net/browse/JDK-8271589): fatal error with variable shift count integer rotate operation.

### Reviewers
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to ef3316b5bf511cfcc2844db002f7f75bcfd4829d
 * [Sandhya Viswanathan](https://openjdk.java.net/census#sviswanathan) (@sviswa7 - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4956/head:pull/4956` \
`$ git checkout pull/4956`

Update a local copy of the PR: \
`$ git checkout pull/4956` \
`$ git pull https://git.openjdk.java.net/jdk pull/4956/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4956`

View PR using the GUI difftool: \
`$ git pr show -t 4956`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4956.diff">https://git.openjdk.java.net/jdk/pull/4956.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/opto/superword.cpp
Comment (vnkozlov) [inline src/hotspot/share/opto/superword.cpp:2491]: Why you removed the assert? What other types can be seen here?
The following code (specifically `in2->get_int()`) will fail if it is not Integer type.

File: src/hotspot/share/opto/vectornode.cpp
Comment (vnkozlov) [inline src/hotspot/share/opto/vectornode.cpp:1163]: This assert is not needed anymore since you already checked for this case.

File: src/hotspot/share/opto/vectornode.cpp
Comment (vnkozlov) [inline src/hotspot/share/opto/vectornode.cpp:1171]: Is this case from Vector API? I think in Java we can have only Int type for shifts.

**Review by vnkozlov — CHANGES_REQUESTED** [review]
I assume `TestLongVectRotate.java` should be also updated.

And you need add cases for Long type vector shifts if they can be generated.

File: src/hotspot/share/opto/vectornode.cpp
Comment (jatin-bhateja) [reply src/hotspot/share/opto/vectornode.cpp:1171]: > Is this case from Vector API? I think in Java we can have only Int type for shifts.
This is to handle long rotates, SLP will insert a ConvI2L before broadcasting shift value into a long vector. In case of constant int value, ConvI2L::Value will create a constant of TypeLong and which is what is being handled here.

File: src/hotspot/share/opto/superword.cpp
Comment (jatin-bhateja) [reply src/hotspot/share/opto/superword.cpp:2491]: The shift is either an int value or a vector of an int value. If it's a vector !in2->is_Con() will make the condition true and no need to check for get_int due to short-circuiting.

File: src/hotspot/share/opto/vectornode.cpp
Comment (jatin-bhateja) [reply src/hotspot/share/opto/vectornode.cpp:1163]: Will remove this before checkin.

**Comment by jatin-bhateja** [comment]
> I assume `TestLongVectRotate.java` should be also updated.
> 
> And you need add cases for Long type vector shifts if they can be generated.

Long rotates patterns with variable shifts i.e. Long.rotateLeft/Right(a[i], int_b[i]/ (int) long_b[i]) are not auto-vectorized currently.
Problem is only seen with rotation over integers.

**Comment by jatin-bhateja** [comment]
Hi @vnkozlov, I have responded to your individual comments, please let me know if there are more comments.

File: src/hotspot/share/opto/superword.cpp
Comment (vnkozlov) [reply src/hotspot/share/opto/superword.cpp:2491]: Okay. IT should be fine because get_int() has own assert.

File: src/hotspot/share/opto/vectornode.cpp
Comment (vnkozlov) [reply src/hotspot/share/opto/vectornode.cpp:1171]: Got it.

**Comment by vnkozlov** [comment]
> > I assume `TestLongVectRotate.java` should be also updated.
> > And you need add cases for Long type vector shifts if they can be generated.
> 
> Long rotates patterns with variable shifts i.e. Long.rotateLeft/Right(a[i], int_b[i]/ (int) long_b[i]) are not auto-vectorized currently.
> Problem is only seen with rotation over integers. I have kept the changes minimal as suggested earlier.

You said in an other comment:
"This is to handle long rotates, SLP will insert a ConvI2L before broadcasting shift value into a long vector."

Is it really a big change for test to cover this case? This is new code in your changes which needs test coverage I think.

I will start testing current changes (with removed assert(is_invariant_vector()) we agreed on).

**Comment by vnkozlov** [comment]
Testing passed clean.

**Comment by jatin-bhateja** [comment]
> > > I assume `TestLongVectRotate.java` should be also updated.
> > > And you need add cases for Long type vector shifts if they can be generated.
> > 
> > 
> > Long rotates patterns with variable shifts i.e. Long.rotateLeft/Right(a[i], int_b[i]/ (int) long_b[i]) are not auto-vectorized currently.
> > Problem is only seen with rotation over integers. I have kept the changes minimal as suggested earlier.
> 
> You said in an other comment:
> "This is to handle long rotates, SLP will insert a ConvI2L before broadcasting shift value into a long vector."
> 
Yes this comment is valid for Long.rotateLeft/Right(a[i], SHIFT/IMM_SHIFT) patterns.

> Is it really a big change for test to cover this case? This is new code in your changes which needs test coverage I think.
> 
> I will start testing current changes (with removed assert(is_invariant_vector()) we agreed on).

**Comment by vnkozlov** [comment]
@jatin-bhateja  I don't understand why you don't want to add testing for `Long.rotateLeft/Right(a[i], SHIFT/IMM_SHIFT) `.
Yes, this bug failed only for Int vectors. But you added code for Longs which have to be verified.

And you need second review.

**Comment by jatin-bhateja** [comment]
> @jatin-bhateja I don't understand why you don't want to add testing for `Long.rotateLeft/Right(a[i], SHIFT/IMM_SHIFT) `.
> Yes, this bug failed only for Int vectors. But you added code for Longs which have to be verified.
> 
> And you need second review.

Hi @vnkozlov ,
My apologies for not mentioning earlier that such test patterns (Long.rotateLeft/Right(a[i], scalar_shift/imm_shift) are already present in jdk/test/hotspot/jtreg/compiler/c2/cr6340864/TestLongVectRotate.java

**Review by vnkozlov — APPROVED** [review]

File: src/hotspot/share/opto/vectornode.cpp
Comment (sviswa7) [inline src/hotspot/share/opto/vectornode.cpp:1174]: This seems redundant. shiftRCnt is being overwritten in the very next statement.

Comment (sviswa7) [reply src/hotspot/share/opto/vectornode.cpp:1174]: Other than this redundant statement, the patch looks good to me.

**Review by sviswa7 — APPROVED** [review]

**Pull request closed without merging**
